File: lesson-2_recursion-and-recurrence/the-longest-path/recursion/main.kt
import kotlin.math.max

fun main() {
    val n = readLine()!!.toInt()
    val t = Array(n) { emptyArray<Int>() }
    for (i in 0 until n) {
        t[i] = Array(i + 1) { 0 }

        val tmp = readLine()!!.split(' ')
        for ((j, v) in tmp.withIndex()) {
            t[i][j] = v.toInt()
        }
    }
    println(getTheLongest(t, n, 0, 0))
}

fun getTheLongest(t: Array<Array<Int>>, n: Int, i: Int, j: Int): Int {
    if (i == n - 1) {
        return t[i][j]
    }
    return max(getTheLongest(t, n, i + 1, j), getTheLongest(t, n, i + 1, j + 1)) + t[i][j]
}

File: lesson-2_recursion-and-recurrence/the-longest-path/recursion/main.mjs
import { stdin as input, stdout as output } from 'node:process'
import * as readline from 'node:readline/promises'

const rl = readline.createInterface({ input, output })

const n = await rl.question('')
const t = Array.from({ length: n }, () => [])
for (let i = 0; i < n; i++) {
    const tmp = (await rl.question('')).split(' ')
    for (const [j, v] of tmp.entries()) {
        t[i][j] = Number(v)
    }
}
console.log(getTheLongest(0, 0))
rl.close()

function getTheLongest(i, j) {
    if (i === n - 1) {
        return t[i][j]
    }
    return Math.max(getTheLongest(i + 1, j), getTheLongest(i + 1, j + 1)) + t[i][j]
}
